lib/Conversion/DCToHW/DCToHW.cpp

```diff
@@ -0,0 +1,946 @@
+  Value bit(Value v, unsigned index, std::optional<StringRef> name = {}) {
+    return extract(v, index, index, name);
+  }
+
+  ///  Creates a hw.array of the given values.
+  Value arrayCreate(ValueRange values, std::optional<StringRef> name = {}) {
+    return buildNamedOp(
+        [&]() { return b.create<hw::ArrayCreateOp>(loc, values); }, name);
+  }
+
+  ///  Extract the 'index'th value from the input array.
+  Value arrayGet(Value array, Value index, std::optional<StringRef> name = {}) {
+    return buildNamedOp(
+        [&]() { return b.create<hw::ArrayGetOp>(loc, array, index); }, name);
+  }
+
+  ///  Muxes a range of values.
+  ///  The select signal is expected to be a decimal value which selects
+  ///  starting from the lowest index of value.
+  Value mux(Value index, ValueRange values,
+            std::optional<StringRef> name = {}) {
+    if (values.size() == 2)
+      return b.create<comb::MuxOp>(loc, index, values[1], values[0]);
+
+    return arrayGet(arrayCreate(values), index, name);
+  }
+
+  ///  Muxes a range of values. The select signal is expected to be a 1-hot
+  ///  encoded value.
+  Value ohMux(Value index, ValueRange inputs) {
+    // Confirm the select input can be a one-hot encoding for the inputs.
+    unsigned numInputs = inputs.size();
+    assert(numInputs == index.getType().getIntOrFloatBitWidth() &&
+           "one-hot select can't mux inputs");
+
+    // Start the mux tree with zero value.
+    // Todo: clean up when handshake supports i0.
+    auto dataType = inputs[0].getType();
+    unsigned width =
+        dataType.isa<NoneType>() ? 0 : dataType.getIntOrFloatBitWidth();
+    Value muxValue = constant(width, 0);
+
+    // Iteratively chain together muxes from the high bit to the low bit.
+    for (size_t i = numInputs - 1; i != 0; --i) {
+      Value input = inputs[i];
+      Value selectBit = bit(index, i);
+      muxValue = mux(selectBit, {muxValue, input});
+    }
+
+    return muxValue;
+  }
+
+  OpBuilder &b;
+  Location loc;
+  Value clk, rst;
+  DenseMap<APInt, Value> constants;
+};
+
+/// Creates a Value that has an assigned zero value. For structs, this
+/// corresponds to assigning zero to each element recursively.
+// NOLINTNEXTLINE(misc-no-recursion)
+static Value createZeroDataConst(RTLBuilder &rtlb, Location loc, Type type) {
+  return TypeSwitch<Type, Value>(type)
+      .Case<NoneType>([&](NoneType) { return rtlb.constant(0, 0); })
+      .Case<IntType, IntegerType>([&](auto type) {
+        return rtlb.constant(type.getIntOrFloatBitWidth(), 0);
+      })
+      .Case<hw::StructType>([&](auto structType) {
+        SmallVector<Value> zeroValues;
+        for (auto field : structType.getElements())
+          zeroValues.push_back(createZeroDataConst(rtlb, loc, field.type));
+        return rtlb.b.create<hw::StructCreateOp>(loc, structType, zeroValues);
+      })
+      .Default([&](Type) -> Value {
+        emitError(loc) << "unsupported type for zero value: " << type;
```

Comment:
NIT: If this should indicate a programmer error, not a user reachable case, `llvm_unreachable` might be better suited.

Reply:
Yeah... emitting an error followed by an assertion is a pattern which generally means you're doing it wrong. Though I do understand why you're doing it.

Generally any `emitError()` not preceeded by a `return` is a sign.